jina/peapods/runtimes/gateway/prefetch.py

```diff
@@ -1,22 +1,57 @@
 import argparse
 import asyncio
-from abc import ABC
-from typing import AsyncGenerator
+from asyncio import Future
+from typing import AsyncGenerator, Dict
 
 from ....helper import typename
 from ....logging.logger import JinaLogger
 from ....types.message import Message
 
-__all__ = ['PrefetchCaller', 'PrefetchMixin']
+__all__ = ['PrefetchCaller']
 
 if False:
     from ...zmq import AsyncZmqlet
 
 
-class PrefetchMixin(ABC):
-    """JinaRPCServicer """
+class PrefetchCaller:
+    """An async zmq request sender to be used in the Gateway"""
 
-    async def Call(self, request_iterator, *args) -> AsyncGenerator[None, Message]:
+    def __init__(self, args: argparse.Namespace, zmqlet: 'AsyncZmqlet'):
+        """
+        :param args: args from CLI
+        :param zmqlet: zeromq object
+        """
+        self.args = args
+        self.zmqlet = zmqlet
+        self.name = args.name or self.__class__.__name__
+
+        self.logger = JinaLogger(self.name, **vars(args))
+        self._message_buffer: Dict[str, Future[Message]] = dict()
+        self._is_running = True
+        self._receive_task = asyncio.create_task(self._receive())
+
+    async def _receive(self):
+        try:
+            while self._is_running:
+                message = await self.zmqlet.recv_message(callback=lambda x: x.response)
+                if message.request_id in self._message_buffer:
+                    self._message_buffer[message.request_id].set_result(message)
+                    del self._message_buffer[message.request_id]
+                else:
+                    self.logger.warning(
+                        f'Discarding unexpected message with request id {message.request_id}'
+                    )
+        except asyncio.CancelledError:
+            pass
+
+    async def close(self):
+        """
+        Stop receiving messages
+        """
+        self._is_running = False
```

Comment:
This is pretty un-forgiving. Should we give the Gateway some seconds after a close, to collect missing messages? Or do we say: You wanted to close it, so you should expect missing messages? (Which would be implemented in the `_receive` function.)

Reply:
I think we should not really wait for messages to come back? We cant really know for how long to wait and if you close this, you probably close the Flow and it should expect incomplete messages?

But we should probably at least cancel also the outstanding futures, which I didnt do